### PR TITLE
Forbid type ascriptions of place expressions in lvalue contexts

### PR DESCRIPTION
**Pull request by b-naber**
Opening to get some feedback.

r? @nikomatsakis 

edit: sorry lcnr, forgot the rust-highfive command

### CODE REVIEW

**Comment by lcnr** [comment]
no worries :+1: this looks impressive for a first pr

r? @nikomatsakis

File: compiler/rustc_typeck/src/check/type_ascription.rs
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/check/type_ascription.rs:39]: This is the right general idea, but I'm worried it's not quite covering all the bases. This will work for something like

```rust
&(x:T)
```

but I don't think it would work for `&(x:T).f`. I'll try to create what seem to me to be a good set of test cases and link them below. 

I'm also not sure if looking at `borrow_expr` is the right thing; the `Place` has a `HirId`, maybe we want to look at that. I have to check under what conditions `borrow_expr` is None.

Comment (nikomatsakis) [reply compiler/rustc_typeck/src/check/type_ascription.rs:39]: Update: OK, I see you added `borrow_expr`. I think instead of doing that, you can use [`expect_expr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.expect_expr) to get the `hir::Expr`. 

To manage the cases I was talking about, you'll need to use a HIR visitor or something to recursively walk the HIR structure.

I am, however, pondering whether the right fix here is to modify the `Place` structure a bit. Let me take a look at that.

Comment (nikomatsakis) [reply compiler/rustc_typeck/src/check/type_ascription.rs:39]: So yes, what I think is that we should modify place construction -- I suspect we should 

* modify [`PlaceBase::Rvalue`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/place/enum.PlaceBase.html#variant.Rvalue) to contain a `HirId`
* treat ascription as an rvalue, during place construction but...probably just everywhere -- that means modifying [this function](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Expr.html#method.is_place_expr) and perhaps [this one](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Expr.html#method.is_syntactic_place_expr), I haven't read that much into that code
* check in this code where `place.base` is an rvalue and, if so, check whether its an ascription and report an error

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Here are some scenarios to test. 

First, we want to test that every kind of borrow errors. Here are the ones I can think of:

* `let ref x = expr: T` should error
* `(expr:T).as_ref()` should error (invoking `AsRef::as_ref` from the standard library, which does auto-ref)
* `&(expr:T)` should error

Next, for each kind of borrow, we want to try both a regular coercion and a coercion that is part of some larger replace. I would say cover these three scenarios:

* `&(expr:T).f` should error
* `&(expr:T)[3]` should error
* `&*(expr:T)`

It's probably enough to cover those three scenarios for *one* kind of borrow, and then test that the other kind of borrows each handle `(expr:T).f` or one of the more complex cases, but ideal would be to test all for all.

File: compiler/rustc_typeck/src/check/type_ascription.rs
Comment (b-naber) [reply compiler/rustc_typeck/src/check/type_ascription.rs:91]: This is fresh code.

File: compiler/rustc_typeck/src/expr_use_visitor.rs
Comment (b-naber) [reply compiler/rustc_typeck/src/expr_use_visitor.rs:48]: I tried to use 'place_with_id' initially, but the `mem_categorization' functionality returns only the place expression of a type ascription in some contexts (or rather in all contexts I tested), so I had to introduce this parameter to get information to decide whether we're dealing with a type ascription. I didn't look into the `mem_categorization` part to see why `diag_expr_id` only corresponds to the place expression part of a type ascription expression.

File: compiler/rustc_typeck/src/check/type_ascription.rs
Comment (b-naber) [reply compiler/rustc_typeck/src/check/type_ascription.rs:39]: See comment above for why I introduced `borrow_expr`. I also tried to somehow (currently don't remember what exactly I tried) to get access to a parent expression (the type ascription expression) for `diag_expr_id`, but wasn't able to do that. 

I will take a look at your suggestions later.

Comment (arora-aman) [reply compiler/rustc_typeck/src/check/type_ascription.rs:39]: I took a look at `mem_categorization` and it isn't easily evident to me why we would get a different expression, maybe I need to spend some more time. The reason the interface includes a `diagnostics_hir_id` is so that we can support providing a hir id that is different from HirId in `PlaceWithHirId` which I think can be utilized here along with `expect_expr`. Assuming that there is a valid reason for `x: T` to not contain the later part of that expression.  (It's currently being only used for destructure patterns).

File: src/test/ui/type-ascription/as_ref.stderr
Comment (b-naber) [inline src/test/ui/type-ascription/as_ref.stderr:34]: I think we can fix these repetitions of errors caused by nested `delegate.borrow` calls using the hir visitor. I want to see whether the general logic is correct now, before doing that.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This is looking pretty good-- I think it's missing a few things, though.

Specifically, I think we want to categorize place expressions as rvalues consistenly. That means we need to edit this function:

https://github.com/rust-lang/rust/blob/399b6452b5d9982438be208668bc758479f13725/compiler/rustc_hir/src/hir.rs#L1478

I was reading into the THIR/MIR construction code to see if any changes are needed there, but I don't *think* so.

File: src/test/ui/coercion/type-ascription/coerce-struct-fields-pass-in-let-stmt.rs
Comment (b-naber) [inline src/test/ui/coercion/type-ascription/coerce-struct-fields-pass-in-let-stmt.rs:24]: There is not error here, but two lines down there is in what looks to be the same expression. Is this correct?

File: compiler/rustc_typeck/src/check/type_ascription.rs
Comment (nikomatsakis) [reply compiler/rustc_typeck/src/check/type_ascription.rs:45]: I am, however, wondering if we really need the finder at all. Do we have an example where it's necessary? I think I expect that if there is a type ascription being borrowed/mutated, it will **always** be the base, since it is an rvalue. I hypothesize that you can just remove this.

**Review by nikomatsakis — APPROVED** [review]
Looking good! Here are some thoughts/comments. Sorry about the delay. Thanks for pushing on this, @b-naber! I am beginning to think we'll be able to get this over the finish line soon.

**Comment by camelid** [comment]
Ping from triage: @b-naber could you address Niko's review? Thanks!

**Comment by b-naber** [comment]
@camelid Landing this will require the lang team to accept the updated RFC we will write for the type ascription feature. I haven't yet addressed Niko's review, because I wanted to wait for that approval. Is that ok?

**Comment by nikomatsakis** [comment]
We could close the PR and re-open it.

**Comment by b-naber** [comment]
OK,  let's do that.

**Comment by camelid** [comment]
> @camelid Landing this will require the lang team to accept the updated RFC we will write for the type ascription feature. I haven't yet addressed Niko's review, because I wanted to wait for that approval. Is that ok?

That's totally fine! I didn't see anything in the PR about waiting on an RFC, so I was confused.

> We could close the PR and re-open it.

So you know for the future: that's not necessary, just indicate that the PR is waiting on something by adding `S-blocked`.

**Pull request closed without merging**
